packages/reakit/src/Form/FormState.ts

```diff
@@ -346,24 +349,34 @@ export function unstable_useFormState<V = Record<any, any>>(
     values: state.values as V,
     validate,
     reset: React.useCallback(() => dispatch({ type: "reset" }), []),
-    submit: React.useCallback(async () => {
-      try {
-        dispatch({ type: "startSubmit" });
-        const validateMessages = await validate();
-        if (onSubmit) {
-          const submitMessages = await onSubmit(state.values as V);
-          const messages = { ...validateMessages, ...submitMessages };
-          dispatch({ type: "endSubmit", messages });
-          if (resetOnSubmitSucceed) {
-            dispatch({ type: "reset" });
-          }
-        } else {
-          dispatch({ type: "endSubmit", messages: validateMessages });
-        }
-      } catch (errors) {
-        dispatch({ type: "endSubmit", errors });
-      }
-    }, [validate]),
+    submit: React.useCallback(
+      () =>
+        Promise.resolve()
+          .then(() => {
+            dispatch({ type: "startSubmit" });
+            return validate();
```

Comment:
Hmm, I think this dispatch should synchronous (outside the Promise).

Reply:
Also, `validate` is a promise. So I guess we can do `validate().then(...)`

Reply:
Both true. Fixed!